### PR TITLE
Convert the particle position pusher to C++

### PR DESCRIPTION
**Pull request by RemiLehe**

### CODE REVIEW

File: Source/Particles/WarpXParticleContainer.cpp [hunk 1]
```diff
@@ -1040,16 +1040,6 @@ WarpXParticleContainer::PushX (int lev, Real dt)
...
 
-            auto& attribs = pti.GetAttribs();
-
-            auto& uxp = attribs[PIdx::ux];
-            auto& uyp = attribs[PIdx::uy];
-            auto& uzp = attribs[PIdx::uz];
-
-            const long np = pti.numParticles();
-
-            giv.resize(np);
```
Comment (RemiLehe) [inline Source/Particles/WarpXParticleContainer.cpp]: No need to allocate `giv` anymore: this is now computed on the fly, inside the kernel.

**Comment by atmyers** [comment]
I suppose we also don't need to copy from AoS to SoA format for the particle positions anymore, either.

**Comment by RemiLehe** [comment]
Oh, interesting! How do we access the particle positions in this case?

**Comment by atmyers** [comment]
You could do something like (not tested):

``` C++
            BL_PROFILE_VAR_START(blp_pp);
            auto& attribs = pti.GetAttribs();
            ParticleType const* AMREX_RESTRICT pstructs = &(pti.GetArrayOfStructs()[0]);
            Real* AMREX_RESTRICT ux = attribs[PIdx::ux].dataPtr();
            Real* AMREX_RESTRICT uy = attribs[PIdx::uy].dataPtr();
            Real* AMREX_RESTRICT uz = attribs[PIdx::uz].dataPtr();
            // Loop over the particles and update the position
            const long np = pti.numParticles();
            const Real inv_c2 = 1./(PhysConst::c*PhysConst::c);
            amrex::ParallelFor( np,
                [=] AMREX_GPU_DEVICE (long i) {
                    // Compute inverse Lorentz factor
                    const Real inv_gamma = 1./std::sqrt(
                        1. + (ux[i]*ux[i] + uy[i]*uy[i] + uz[i]*uz[i])*inv_c2);
                    // Update positions over one time step
                    pstructs[i].pos(0) += ux[i] * inv_gamma * dt;
#if (AMREX_SPACEDIM == 3) || (defined WARPX_RZ) // RZ pushes particles in 3D
                    pstructs[i].pos(1) += uy[i] * inv_gamma * dt;
#endif
                    pstructs[i].pos(2) += uz[i] * inv_gamma * dt;
                }
            );
BL_PROFILE_VAR_STOP(blp_pp);
```

**Comment by RemiLehe** [comment]
OK, I'll try it.
Another question: my understanding is that any `BL_PROFILE_VAR_START/STOP` will be meaningless if it is *inside* the particle tile loop, when running on GPU (because there is only CPU-GPU synchronization at the end of this loop). Is this correct?

**Comment by atmyers** [comment]
Yes, that's correct.

**Comment by RemiLehe** [comment]
Ok, so as part of this PR, I guess I should also move the `BL_PROFILE` outside of the loop.

**Comment by dpgrote** [comment]
The way that the particle push is being done here looks wrong for 2D and completely breaks the RZ. In 2D, there should be only pstructs[i].pos(0) and pstructs[i].pos(1). For RZ, the pos(0) will be the radius and pos(1) the z position. The original code was relying on GetPosition to return the x, y, z arrays which are calculated from pos(0), pos(1), and the extra theta component. The pos cannot be used directly.

Also, PushX is only used by the ES version of the code. The EM uses the routine PushPX (which calls warpx_particle_pusher.

**Comment by atmyers** [comment]
Thanks @dpgrote, you are right and I should have caught that. 

I think that for the C++ version of these routines, we should have a AMREX_GPU_HOST_DEVICE function that gets the x, y, and z position for a single particle, accounting for RZ, so we don't need the temporary arrays any more.

**Comment by atmyers** [comment]
It actually looks to me like PushX isn't used at all - there is a separate PushXES routine for the electrostatic update.

**Comment by RemiLehe** [comment]
OK, my bad!
I'll do a separate PR then where:
- I'll remove `PushX` (and check that it is not used, i.e. that nothing breaks)
- I'll modify `PushPX` and `PushES`

**Pull request merged**